### ISSUE TITLE
@defer: support in Apple and JS HttpEngines

### ISSUE THREAD

**Issue by BoD** [apollographql/apollo-kotlin#3982]
When using `@defer` the expectation is that the deferred parts are surfaced as soon as they arrive.

However the current implementations of the [JS](https://github.com/apollographql/apollo-kotlin/blob/c6a16054c57e9c3b2c2c93f895d3169b37186d8a/apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/http/KtorHttpEngine.kt#L49) and [Apple](https://github.com/apollographql/apollo-kotlin/blob/584d5a5ae94e6deb13568552bfa8d19f389ed588/apollo-runtime/src/appleMain/kotlin/com/apollographql/apollo3/network/http/NSURLSessionHttpEngine.kt#L135) of `DefaultHttpEngine` buffer the whole response before returning it, making all the parts emitted only after the last one has been received.

**Comment by BoD** [comment]
- Apple implementation is ✅ in a specific `HttpEngine` (caveat: it requires the new memory model), in #4073
- JS implementation is based on KTOR, which JS implementation currently doesn't support streaming, even though the generic API allows it. See [this issue](https://youtrack.jetbrains.com/issue/KTOR-4274/HttpClient-Streaming-support-in-JS).

**Comment by jpvajda** [comment]
@martinbonnin I noticed this issue in Todo for `@defer` Will we make this change as a follow up release?

**Comment by martinbonnin** [comment]
@jpvajda ideally we'd wait for KTOR to implement the feature before adding it in Apollo Kotlin. If it becomes too much of an issue for us, we could consider contributing the fix there.

**Comment by jpvajda** [comment]
@martinbonnin sounds good! we can discuss what release to put this in in our sync today.

**Comment by jpvajda** [comment]
We discussed this today and we'll wait to see the resolution on this dependency https://youtrack.jetbrains.com/issue/KTOR-4274/HttpClient-Streaming-support-in-JS.
